### ISSUE TITLE
Error message pointing to struct definition instead of pointing to trait impl with missing trait bound

### ISSUE THREAD

**Issue by brunoczim** [rust-lang/rust#93684]
I tried this code:

```rust
trait TraitFoo {
    type Bar;
}

struct Foo<T>
where
    T: TraitFoo,
{
    inner: T::Bar,
}

impl<T> Clone for Foo<T>
where
    T: TraitFoo,
    T::Bar: Clone,
{
    fn clone(&self) -> Self {
        Self { inner: self.inner.clone() }
    }
}

impl<T> Copy for Foo<T> {}
```

I expected to see this happen: an error on line 22 about a missing trait bound

Instead, this happened: a confusing error message on line 9

Message:
```
error[[E0277]](https://doc.rust-lang.org/nightly/error-index.html#E0277): the trait bound `T: TraitFoo` is not satisfied
 [--> src/lib.rs:9:5
](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=8c8a001e736ea8d9335cfead8c7aaadd#)  |
9 |     inner: T::Bar,
  |     ^^^^^^^^^^^^^ the trait `TraitFoo` is not implemented for `T`

For more information about this error, try `rustc --explain E0277`.
error: could not compile `playground` due to previous error
```

### Meta
`rustc --version --verbose`:
```
rustc 1.60.0-nightly (71226d717 2022-02-04)
```
Tested on both stable and nightly tho.

Backtrace is not applicable here, since the bug is a confusing error message.

### EDIT
Fixing the code would result in this:
```rust
trait TraitFoo {
    type Bar;
}

struct Foo<T>
where
    T: TraitFoo,
{
    inner: T::Bar,
}

impl<T> Clone for Foo<T>
where
    T: TraitFoo,
    T::Bar: Clone,
{
    fn clone(&self) -> Self {
        Self { inner: self.inner.clone() }
    }
}

impl<T> Copy for Foo<T>
where
    T: TraitFoo,
    T::Bar: Copy,
{}
```

**Comment by brunoczim** [comment]
Closing this as I used the wrong template
